locator/util.cc

```diff
@@ -98,35 +98,38 @@ get_range_to_address_map(locator::effective_replication_map_ptr erm) {
 future<utils::chunked_vector<dht::token_range_endpoints>>
 describe_ring(const replica::database& db, const gms::gossiper& gossiper, const sstring& keyspace, bool include_only_local_dc) {
     utils::chunked_vector<dht::token_range_endpoints> ranges;
+    std::unordered_map<host_id, describe_ring_endpoint_info> host_infos;
 
     auto erm = db.find_keyspace(keyspace).get_vnode_effective_replication_map();
     std::unordered_map<dht::token_range, host_id_vector_replica_set> range_to_address_map = co_await (
             include_only_local_dc
                     ? get_range_to_address_map_in_local_dc(erm)
                     : get_range_to_address_map(erm)
     );
+    ranges.reserve(range_to_address_map.size());
     auto tmptr = erm->get_token_metadata_ptr();
-    for (auto entry : range_to_address_map) {
-        const auto& topology = tmptr->get_topology();
-        auto range = entry.first;
-        auto addresses = entry.second;
+    const auto& topology = tmptr->get_topology();
+    for (const auto& [range, addresses] : range_to_address_map) {
         dht::token_range_endpoints tr;
         if (range.start()) {
             tr._start_token = range.start()->value().to_sstring();
         }
         if (range.end()) {
             tr._end_token = range.end()->value().to_sstring();
         }
+        tr._endpoints.reserve(addresses.size());
+        tr._rpc_endpoints.reserve(addresses.size());
+        tr._endpoint_details.reserve(addresses.size());
         for (auto endpoint : addresses) {
-            dht::endpoint_details details;
-            details._host = gossiper.get_address_map().get(endpoint);
-            details._datacenter = topology.get_datacenter(endpoint);
-            details._rack = topology.get_rack(endpoint);
-            tr._rpc_endpoints.push_back(gossiper.get_rpc_address(endpoint));
-            tr._endpoints.push_back(fmt::to_string(details._host));
-            tr._endpoint_details.push_back(details);
+            auto it = host_infos.find(endpoint);
+            if (it == host_infos.end()) {
+                it = host_infos.emplace(endpoint, get_describe_ring_endpoint_info(endpoint, topology, gossiper)).first;
```

Comment:
Why is get_describe_ring_endpoint_info() so slow?

Reply:
Do you ask why it needed the optimization, or is there an indication it is so slow with this change?

Reply:
It just did a map lookup for every member:
- `gossiper.get_address_map().get(endpoint)`
- `topology.get_datacenter(endpoint)`
- `topology.get_rack(endpoint)`
- `gossiper.get_rpc_address(endpoint)` (that has a nested map lookup: first for the endpoint_state_ptr then for the RPC_ADDRESS state)

Reply:
This summed up to 20% reduction in runtime when idle.
The reason to optimize it at all is that it was observed to take a long time (over 30 seconds) in the field during repair (that also runs in the streaming scheduling group) and presumingly under query load as well.

Reply:
So we're optimizing from over 30 seconds to over 24 seconds?

It's a nice improvement, but does it doesn't materially change anything, it's still to long for human interaction.

What's the cpu usage of the thing? Maybe the problem was scheduling, not runtime (and then the solution is to provide a separate nested scheduling groups for the data plane and for the control plan under streaming).

Reply:
Then I don't understand what we're doing here. 

If we're optimizing by 20%, then whatever problem caused it to go over 30 seconds could cause it to go over 36 seconds and we're back where we started.